packages/ensnode-sdk/src/api/name-tokens/response.ts

```
@@ -0,0 +1,137 @@
+ */
+export type NameTokensResponseCode =
+  (typeof NameTokensResponseCodes)[keyof typeof NameTokensResponseCodes];
+
+/**
+ * Error codes for Name Tokens API responses with 'error' response code.
+ */
+export const NameTokensResponseErrorCodes = {
+  /**
+   * Represents a response error for the unknown name context error.
+   */
+  UnknownNameContext: "unknown-name-context",
+
+  /**
+   * Represents a response error for the generic error.
+   */
+  Generic: "generic",
+} as const;
+
+/**
+ * The derived string union of possible {@link NameTokensResponseErrorCodes}.
+ */
+export type NameTokensResponseErrorCode =
+  (typeof NameTokensResponseErrorCodes)[keyof typeof NameTokensResponseErrorCodes];
+
+/**
+ * Token Statuses
+ *
+ * Extends {@link NFTMintStatuses} with additional "virtual" statuses.
+ */
+export const RegisteredNameTokenStatuses = {
+  ...NFTMintStatuses,
+
+  Expired: "expired",
+} as const;
+
+/**
+ * Token Status
+ *
+ * Extends {@link NFTMintStatus} with additional "virtual" statuses.
+ */
+export type RegisteredNameTokenStatus =
+  (typeof RegisteredNameTokenStatuses)[keyof typeof RegisteredNameTokenStatuses];
+
+export interface RegisteredNameToken {
+  token: NameToken;
+
+  /**
+   * Registration Lifecycle
+   *
+   * The latest Registration Lifecycle for a node referenced in
+   * `token.domainAsset.domainId`.
+   */
+  registrationLifecycle: RegistrationLifecycle;
+
+  /**
+   * Token Status
+   *
+   * Derived from `token.mintStatus`.
+   *
+   * Set to `expired` when:
+   * - the `token.mintStatus` is "mint" however the latest registration lifecycle expiresAt is in the past.
+   */
+  tokenStatus: RegisteredNameTokenStatus;
+}
+
+/**
+ * A response when Name Tokens API can respond with requested data.
+ */
+export type NameTokensResponseOk = {
+  responseCode: typeof NameTokensResponseCodes.Ok;
+
+  /**
+   * Name Tokens for the requested name.
+   */
+  nameTokens: RegisteredNameToken[];
```

It will be high value to document some ideas here. We need to explicitly document the potential state space of responses. Suggest to look for names with each of the state permutations I cite below and then consider creating some integration test for the API responses we make for them. Advice appreciated.

1. What does this array contain? -> It contains EVERY tokenized representation of the name that has ever been indexed for the given name as of `accurateAsOf`, even if the given token has been burned or expired.
2. What does it mean if this array is empty? -> It means the name has never been registered / minted before.
3. What does it mean if this array contains more than 1 item? ->
   1. It means that more than 1 distinct tokenized representation of the ownership of the name has been indexed for the given name as of `accurateAsOf`.
   2. All possible permutations of mint statuses of these tokens is possible. Multiple could be actively minted. Multiple could be burnt. Some could be burnt, others could be minted.
   3. We should give specific guidance for how direct subnames of .eth that are wrapped by the NameWrapper will be represented.
      1. Case A: A direct subname of .eth that has never been registered.
      2. Case B: A direct subname of .eth that has been registered but has never been wrapped by the NameWrapper, and...:
         1. Case B.1: Is still actively minted (independent of its expiry state)
         2. Case B.2: Has been burned by sending it to the null address.
      5. Case C: A direct subname of .eth that has been registered and has been wrapped by the NameWrapper, and...
         1. Case C.1: Is still actively wrapped by the NameWrapper (independent of its expiry state)
         2. Case C.2: Is no longer wrapped by the NameWrapper, but is still actively minted by the BaseRegistrar (independent of its expiry state)
         3. Case C.3: Is no longer wrapped by the NameWrapper, and is also no longer minted by the BaseRegistrar (both tokens now burned by sending to the null address)
4. In what ordering will the values in the array be returned? Please suggest what we should use.
5. Please cite what is guaranteed to be distinct about each value in the array.
6. Explicitly document why it can be useful to get tokenized representations of the name that are now burnt: This can be helpful for looking up historical activity for the name, including past buy orders, sell orders, and sales.

---

@lightwalker-eth I suggest we don't include the case when `nameTokens` array is empty. We simply should return a 404 error with explanation how no name tokens were never indexed for the requested name / domainId. It will simplify the data model (we won't have to wonder what to do with `expiresAt` field for a name that was never indexed 😅).